presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,392 @@
+        private boolean isConnected(Set<PlanNode> nodes)
+        {
+            checkState(!nodes.isEmpty(), "nodes cannot be empty");
+            return getConnectedNodeSets(nodes).size() == 1;
+        }
+
+        private Set<Set<PlanNode>> getConnectedNodeSets(Collection<PlanNode> nodes)
+        {
+            Set<Set<PlanNode>> nodeSets = new HashSet<>();
+            for (PlanNode node : nodes) {
+                Set<Set<PlanNode>> newNodeSets = new HashSet<>();
+                ImmutableSet.Builder<PlanNode> newNodeSet = ImmutableSet.builder();
+                newNodeSet.add(node);
+
+                if (edges.containsKey(node)) {
+                    Set<PlanNode> connectedNodes = edges.get(node);
+                    for (Set<PlanNode> nodeSet : nodeSets) {
+                        if (intersection(nodeSet, connectedNodes).isEmpty()) {
+                            newNodeSets.add(nodeSet);
+                        }
+                        else {
+                            newNodeSet.addAll(nodeSet);
+                        }
+                    }
+                }
+                else {
+                    newNodeSets.addAll(nodeSets);
+                }
+                newNodeSets.add(newNodeSet.build());
+                nodeSets = newNodeSets;
+            }
+
+            return nodeSets;
+        }
+
+        private static List<JoinNode.EquiJoinClause> flipJoinCriteriaIfNecessary(List<JoinNode.EquiJoinClause> joinCriteria, PlanNode left)
+        {
+            return joinCriteria
+                    .stream()
+                    .map(criterion -> left.getOutputSymbols().contains(criterion.getLeft()) ? criterion : criterion.flip())
+                    .collect(toImmutableList());
+        }
+
+        private static PlanNode getJoinSource(PlanNodeIdAllocator idAllocator, List<PlanNode> nodes, Expression filter, List<Symbol> outputSymbols)
+        {
+            PlanNode planNode;
+            if (nodes.size() == 1) {
+                planNode = getOnlyElement(nodes);
+                if (!(BooleanLiteral.TRUE_LITERAL).equals(filter)) {
+                    return new FilterNode(idAllocator.getNextId(), planNode, filter);
+                }
+                return planNode;
+            }
+            return new JoinGraphNode(idAllocator.getNextId(), nodes, filter, outputSymbols);
+        }
+
+        private JoinNode chooseJoinOrder(JoinGraphNode joinGraph)
+        {
+            JoinNode join = memo.get(joinGraph);
+            if (join == null) {
+                join = generatePartitions(joinGraph.getSources().size())
+                        .map(partitioning -> createJoinNodeTree(joinGraph, partitioning, idAllocator))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .min(planNodeOrdering)
+                        .orElseThrow(() -> new IllegalStateException("joinOrders cannot be empty"));
+                memo.put(joinGraph, join);
```

Comment:
Would it be useful to log the minimum join order chosen and its cost here for debugging?